api/server.go

```diff
@@ -0,0 +1,408 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	task "github.com/cedana/cedana/api/services/task"
+	"github.com/cedana/cedana/container"
+	"github.com/cedana/cedana/utils"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+)
+
+// Unused for now...
+type GrpcService interface {
+	Register(*grpc.Server) error
+}
+
+type UploadResponse struct {
+	UploadID  string `json:"upload_id"`
+	PartSize  int    `json:"part_size"`
+	PartCount int    `json:"part_count"`
+}
+
+type service struct {
+	Client            *Client
+	ClientLogStream   task.TaskService_LogStreamingServer
+	ClientStateStream task.TaskService_ClientStateStreamingServer
+	r                 *os.File
+	w                 *os.File
+	task.UnimplementedTaskServiceServer
+}
+
+func (s *service) Dump(ctx context.Context, args *task.DumpArgs) (*task.DumpResp, error) {
+
+	// Close before dumping
+	s.r.Close()
+	s.w.Close()
+	// var zipFileSize uint64
+	cfg := utils.Config{}
+	store := utils.NewCedanaStore(&cfg)
+	s.Client.Process = &task.ProcessInfo{}
+
+	s.Client.Process.PID = args.PID
+
+	if args.Type != task.DumpArgs_MARKET {
+		var state task.ProcessState
+
+		state.Flag = task.FlagEnum_JOB_RUNNING
+		state.PID = args.PID
+
+		err := s.Client.db.CreateOrUpdateCedanaProcess("NOT_MARKET", &state)
+		if err != nil {
+			return nil, err
+		}
+
+	}
+
+	err := s.Client.Dump(args.Dir, args.PID)
+	if err != nil {
+		return nil, err
+	}
+
+	switch args.Type {
+	case task.DumpArgs_SELF_SERVE:
+		// TODO BS: This is a hack for now
+	case task.DumpArgs_MARKET:
+		file, err := os.Open(s.Client.CheckpointDir + ".zip")
```

Comment:
CheckpointDir should go and be replaced by queries to the db for the checkpoint, makes this interface a lot cleaner